File: src/test/java/it/xpug/kata/birthday_greetings/AcceptanceTest.java
package it.xpug.kata.birthday_greetings;

import it.xpug.kata.birthday_greetings.employee.EmployeeRepository;
import it.xpug.kata.birthday_greetings.messaging.MessagingService;
import it.xpug.kata.birthday_greetings.template.HappyBirthdayTemplate;
import org.junit.Before;
import org.junit.Test;

import java.text.ParseException;

import static java.util.Arrays.asList;
import static org.mockito.Mockito.*;


public class AcceptanceTest {

    private BirthdayService birthdayService;
    private MessagingService messagingService = mock(MessagingService.class);
    private HappyBirthdayTemplate expectedHappyBirthdayTemplate;
    private EmployeeRepository employeeRepository = mock(EmployeeRepository.class);
    private Employee employee;

    @Before
    public void setUp() throws ParseException {
        employee = new Employee("John", "Doe", "1982/10/08", "john.doe@foobar.com");
        birthdayService = new BirthdayService(messagingService, employeeRepository);
        expectedHappyBirthdayTemplate = new HappyBirthdayTemplate(employee);
    }

    @Test
    public void willSendGreetings_whenItsSomebodysBirthday() throws Exception {

        when(employeeRepository.load()).thenReturn(asList(employee));

        birthdayService.sendGreetings(new XDate("2008/10/08"));
        verify(messagingService).sendMessage("sender@here.com", expectedHappyBirthdayTemplate);
    }

    @Test
    public void willNotSendEmailsWhenNobodysBirthday() throws Exception {
        when(employeeRepository.load()).thenReturn(asList(employee));

        birthdayService.sendGreetings(new XDate("2008/01/01"));
        verifyNoMoreInteractions(messagingService);
    }
}

File: src/test/kotlin/it/xpug/kata/birthday_greetings/AcceptanceTest.kt
package it.xpug.kata.birthday_greetings

import org.junit.Assert.*

import it.xpug.kata.birthday_greetings.conf.Config
import org.junit.*

import com.dumbster.smtp.*
import org.junit.runner.RunWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.junit4.SpringRunner

import javax.mail.Session

@RunWith(SpringRunner::class)
@SpringBootTest(classes = arrayOf(Config::class))
@ActiveProfiles("test")
class AcceptanceTest {
    private val mailServer = SimpleSmtpServer.start(NONSTANDARD_PORT)

    @Autowired
    lateinit private var birthdayService: BirthdayService

    @After
    fun tearDown() {
        mailServer.stop()
        Thread.sleep(200)
    }

    @Test
    fun willSendGreetings_whenItsSomebodysBirthday() {

        birthdayService.sendGreetings(XDate("2008/10/08"))

        assertEquals("message not sent?", 1, mailServer!!.receivedEmailSize.toLong())
        val message = mailServer!!.receivedEmail.next() as SmtpMessage
        assertEquals("Happy Birthday, dear John!", message.body)
        assertEquals("Happy Birthday!", message.getHeaderValue("Subject"))
        val recipients = message.getHeaderValues("To")
        assertEquals(1, recipients.size.toLong())
        assertEquals("john.doe@foobar.com", recipients[0].toString())
    }

    @Test
    fun willNotSendEmailsWhenNobodysBirthday() {
        birthdayService.sendGreetings(XDate("2008/01/01"))
        assertEquals("what? messages?", 0, mailServer!!.receivedEmailSize.toLong())
    }

    companion object {
        private val NONSTANDARD_PORT = 9999
    }
}
